### PR TITLE
Add support and docs for an auto-publish based local dev workflow.

### LINKED ISSUE #2691: Streamline the appservices+fenix local development workflow
Our docs currently recommend two ways of building consuming apps against local changes to this repo: an automated-but-perpetually-buggy composite build workflow, and a reliable-but-tedious manual workflow of publishing to a local maven repo.  It's my understanding that we all currently use them manual local publication workflow because at least it works reliably.

It should not be so painful to make a local build of Fenix using a locally-modified copy of appservices. This issue tracks improving that situation. We can close it when there is a reliable way to make a local build of Fenix+appservices+android-components with a single build command.

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/SYNC-857)
┆Story Points: 5
┆Epic: Automation and tooling to ensure that downstream integration of our releases is smooth and reliable
┆Sprint: SYNC - end 2020-03-13

### PR DESCRIPTION
**Pull request by rfk**
Our docs currently recommend two ways of building consuming apps against
local changes to this repo: an automated-but-perpetually-buggy composite build
workflow, and a reliable-but-tedious manual workflow of publishing to a local
maven repo.

This commit removes them both and replaces them with something similar to
the workflow used by android-components: some scripting to automate publishing
to and consuming from a local maven repo.

Fixes #2691.

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - `automation/all_tests.sh` runs to completion and produces no failures
  - Note: For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/master/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

**Comment by rfk** [comment]
@grigoryk IIRC you added the autoPublish workflow for a-c to Fenix, have you been happy with how it's worked out in practice? What do you think about doing something similar for appservices?  (I'd be happy to work on a consumer-side PR for a-c if this seems like a good idea).

**Comment by rfk** [comment]
See https://github.com/mozilla-mobile/fenix/pull/8681 for the corresponding consumer-side changes, using Fenix as a testcase.

File: build-scripts/substitute-local-appservices.gradle [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+                all { dependency ->
+                    if (!(dependency.requested instanceof ModuleComponentSelector)) {
+                        // We only care about substituting for a module, not a project.
+                        return
+                    }
+
+                    // For every org.mozilla.appservices.* module, substitute its version for '+'.
+                    // '+' version tells gradle to resolve the latest available version.
+                    // As long as 'mavenLocal' is in the repositories list, gradle should pick out
+                    // latest published module during dependency resolution phase.
```
Comment (rfk) [inline build-scripts/substitute-local-appservices.gradle]: Hrm, this doesn't seem to work unless I increment the version number in `.build-config.yml`. If the latest published version if `0.52.0` and my local publishing workflow creates a `0.52.0-LOCAL-TIMESTAMP` then gradle will prefer the `0.52.0` version. Maybe I need to tweak the format of the version numbers generated locally.

Comment (rfk) [reply build-scripts/substitute-local-appservices.gradle]: Going to hack around this by forcing the locally-published version number to be astronomically high.

**Comment by rfk** [comment]
> @grigoryk IIRC you added the autoPublish workflow for a-c to Fenix,
> have you been happy with how it's worked out in practice?

From out-of-band conversations: yes, he's been happy with it overall.

**Comment by rfk** [comment]
Here's a corresponding a-c change: https://github.com/mozilla-mobile/android-components/pull/6218/

**Comment by rfk** [comment]
I was going to demo this in the project meeting today, but couldn't figure out how to do it without making everyone wait for a bunch of rust and java code to recompile. Though the magic of computers I made an async demo with the "waiting for things to compile" edited out:

  https://www.youtube.com/watch?v=n1Gt3Zt_xBI

File: build.gradle [hunk 2]
```diff
@@ -277,3 +277,15 @@ task ktlintFormat(type: JavaExec, group: "formatting") {
...
+// Extremely unsophisticated way to publish a local development version while hiding implementation details.
+//
+// For now we just run `publishToMavenLocal` while setting the special property `local`, and have some logic
+// in our gradle scripts to construct an auto-incrementing version number when this property is set.
+//
+// I hope to grow this into something more sophisticated over time (e.g. to avoid re-publishing when nothing
+// has changed, perhaps cleanup old versions, etc) while keeping a stable `./gradlew autoPublishForLocalDevelopment`
+// interface  for consumers.
+task autoPublishForLocalDevelopment(type: Exec) {
+  commandLine "./gradlew", "publishToMavenLocal", "-Plocal=true"
```
Comment (grigoryk) [inline build.gradle]: Interestingly, but even though in theory, iiuc, you should just be able to do something like `dependsOn('publishToMavenLocal')`, in practice gradle doesn't seem to consider `publish*` tasks as part of the larger task graph...

Comment (rfk) [reply build.gradle]: I think the current approach with the `local=true` property may also require that to be available pretty early in the configuration phase, possibly preceding the construction of the task graph.

**Comment by rfk** [comment]
This seems to be providing some value in practice, so I'm going to go ahead and land once CI gives me the all-clear, and we can iterate from there.

**Pull request merged**